### PR TITLE
[Swift] Add functions concept exercise

### PR DESCRIPTION
**Pull request by wneumann**

### CODE REVIEW

File: languages/swift/exercises/concept/arrays/.docs/after.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [Arrays][array] are one of Swift's three primary collection types. Arrays are ordered lists of elements where the elements can be of any type, however, all elements of any given list must have the same type.
 
-Arrays literals are written as a series of elements, separated by commas, enclosed in square brackets. Empty arrays are just a pair of square brackets. Type names for arrays are written in one of two ways: `Array<T>` or `[T]` where `T` is the type of the elements in thee array. When creating an empty array, the type must be specified.
+Arrays literals are written as a series of elements, each separated by commas, enclosed in square brackets. Empty arrays are just a pair of square brackets. Type names for arrays are written in one of two ways: `Array<T>` or `[T]` where `T` is the type of the elements in thee array. When creating an empty array, the type must be specified.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/arrays/.docs/after.md:3]: @wneumann This change should probably not be part of this PR, but instead be a separate PR.

File: languages/swift/exercises/concept/arrays/.docs/introduction.md [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [Arrays][array] are one of Swift's three primary collection types. Arrays are ordered lists of elements where the elements can be of any type, however, all elements of any given list must have the same type.
 
-Arrays literals are written as a series of elements, separated by commas, enclosed in square brackets. Empty arrays are just a pair of square brackets. Type names for arrays are written in one of two ways: `Array<T>` or `[T]` where `T` is the type of the elements in thee array. When creating an empty array, the type must be specified.
+Arrays literals are written as a series of elements, each separated by commas, enclosed in square brackets. Empty arrays are just a pair of square brackets. Type names for arrays are written in one of two ways: `Array<T>` or `[T]` where `T` is the type of the elements in thee array. When creating an empty array, the type must be specified.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/arrays/.docs/introduction.md:3]: @wneumann This change should probably not be part of this PR, but instead be a separate PR.

File: languages/swift/exercises/concept/functions/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+```swift
+quantities(layers: "sauce", "noodles", "sauce", "meat", "mozzarella", "noodles", "sauce", "eggplant", "béchamel", "noodles", "sauce", "mozzarella")
+// => (noodles 9, sauce 0.8)
+quantities(layers: "sauce", "noodles", "sauce", "noodles", "meat", "noodles", "mozzarella")
+// => (noodles 9, sauce 0.4)
+```
+
+## 4. In-out parameters
+
+Unfortunately, the sauce you have on hand was made in America, and the volume is listed in fluid ounces instead of liters. To help let you know if you have enough sauce on hand, write a function, `toOz`, that takes an in-out tuple parameter, `(noodles: Int, sauce: Double)`, and converts the quantity of sauce to fluid ounces. Note that there are 33.814 fluid ounces per liter.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/README.md:46]: Just to prevent people from reading "Unfortunately, the sauce you have on hand was made in America" as something negative about America:

```suggestion
Unfortunately, the sauce you have on hand has the volume is listed in fluid ounces instead of liters. To help let you know if you have enough sauce on hand, write a function, `toOz`, that takes an in-out tuple parameter, `(noodles: Int, sauce: Double)`, and converts the quantity of sauce to fluid ounces. Note that there are 33.814 fluid ounces per liter.
```

File: languages/swift/exercises/concept/functions/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+// => 10
+remainingMinutesInOven(elapsedMinutes: 30, expectedMinutesInOven: 75)
+// => 45
+```
+
+## 2. Variadic parameters
+
+You had previously written a `preparationTimeInMinutes` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare. However, you are not particularly happy with this function, as you have to count the number of layers in your lasagna yourself before you can call this function, and who has time for all that?
+
+To make things easier for yourself, write a function, `preparationTimeInMinutes` that takes a variadic String parameter, `layers`, where you list the different layers in your lasagna. The function will then count the number of layers in your lasagna, and return the amount of preparation time required.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/README.md:22]:
```suggestion
To make things easier for yourself, write a function, `preparationTimeInMinutes` that takes a variadic `String` parameter, `layers`, where you list the different layers in your lasagna. The function will then count the number of layers in your lasagna, and return the amount of preparation time required.
```

File: languages/swift/exercises/concept/functions/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
+In this exercise you're going to write some more code to help you cook your brilliant lasagna from your favorite cooking book.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/README.md:1]: I really like how you've expanded on the `basics` exercise's story in the tasks!

File: languages/swift/exercises/concept/functions/.docs/hints.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+## 1. Default parameter values
+
+- [Default parameter values][default-parameter-values] can be specified by adding `= value` after the parameter's type annotation. Where _value_ is the default you want to use
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/.docs/hints.md:3]: This hint gives away the solution in its text. Could you perhaps rephrase to not directly give away the solution?

This comment also applies to some of the other hints.

File: languages/swift/exercises/concept/functions/.docs/hints.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+## 1. Default parameter values
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/.docs/hints.md:1]: The hints are very useful! 👍

File: languages/swift/exercises/concept/functions/.docs/introduction.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    for char in "abcdefghijklmnopqrstuvwxyz" {
+      if char.isLetter && !base.contains(char) {
+        missing.append(char)
+      }
+    }
+    print("missing:", missing)
+    return missing
+  }
+  return "\(str) I say, \"\(missingLetters())\""
+}
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/.docs/introduction.md:110]: This example is quite big, which can make it somewhat harder for the student to quickly grasp what it happening. Maybe you could come up with a simpler example?

File: languages/swift/exercises/concept/functions/.docs/after.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+Great job! Functions are at the heart of Swift programming, so understanding them is vital to becoming fluent in Swift.
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/.docs/after.md:1]: The `after.md` file contains less examples than the `introduction.md` file, but that should probably be the other way around.

File: languages/swift/exercises/concept/functions/.docs/after.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+## Key Takeaways from this Exercise
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/.docs/after.md:3]: This "Key Takeaways" format is different from the other Swift exercise's `after.md` files. Is that intentional? It might be worth standardizing on one format.

File: languages/swift/exercises/concept/functions/Sources/functions/functions.swift [hunk 11]
```diff
@@ -0,0 +1,9 @@
+// Implement function remainingMinutesInOven
+
+// Implement function preparationTimeInMinutes
+
+// Implement function quantities
+
+// Implement function toOz
+
+// Implement function redWine
```
Comment (ErikSchierboom) [inline languages/swift/exercises/concept/functions/Sources/functions/functions.swift:9]: For consistency with the `basics` exercise:

```suggestion
// TODO: define the 'remainingMinutesInOven' function

// TODO: define the 'preparationTimeInMinutes' function

// TODO: define the 'quantities' function

// TODO: define the 'toOz' function

// TODO: define the 'redWine' function
```

Note that usually we define stubs for the functions to implement, but as this exercise is about functions themselves, I think TODO's work very well here.

File: languages/swift/exercises/concept/functions/.docs/introduction.md (hunk 8 above)
Comment (wneumann) [reply languages/swift/exercises/concept/functions/.docs/introduction.md:110]: Let me see what I can come up with.

File: languages/swift/exercises/concept/functions/.docs/after.md (hunk 10 above)
Comment (wneumann) [reply languages/swift/exercises/concept/functions/.docs/after.md:3]: Yes, this was a question that had arisen earlier. There was some concern that the after looking so similar to the introduction might lead the student to skim it and miss the new information. I had seen this style of after in IIRC the Elixir track and wanted other eyeballs to look at it and try to see which was better. I personally prefer the style that is similar to the introduction.md.

Comment (ErikSchierboom) [reply languages/swift/exercises/concept/functions/.docs/after.md]: Yes me too. It is also worth to note that we will make it so that students will probably want to revisit these after.md files later on, as a sort of reference, at which point they'll probably don't even remember the introduction.md document.
